File: Day 38/solution.java
class Solution {
  public int rob(int[] nums) {
    Map<Integer, Integer> maxMoneyForRob = new HashMap<>();
    for (int i = 0; i < nums.length; ++i) {
      int curr = maxMoneyForRob.getOrDefault(i-2, 0) + nums[i];
      maxMoneyForRob.put(i, Math.max(curr, maxMoneyForRob.getOrDefault(i-1, 0)));
    }
    return maxMoneyForRob.get(nums.length-1);
  }
}

File: Day 38/solution.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [-1 for i in range(n + 2)]
        return self.recur(nums, dp, n - 1)

        
    def recur(self, nums, dp, n) -> int:
        if n < 0:
            return 0
        elif dp[n] >= 0:
            return dp[n]
        ans = max(self.recur(nums, dp, n - 2) + nums[n], self.recur(nums, dp, n - 1))
        dp[n] = ans
        return dp[n]